### ISSUE TITLE
Feature Ruest: Keep last N docker images by laste-download date

### ISSUE THREAD

**Issue by tahajahangir** [devopshq/artifactory-cleanup#61]
It's reasonable to keep N last downloaded (or N last created) tags per image.

**Comment by donhui** [comment]
The rule `KeepLatestNFilesInFolder(count=N)` may be useful for you.

**Comment by allburov** [comment]
@tahajahangir did KeepLatestNFilesInFolder help you?

**Comment by tahajahangir** [comment]
No, the file-based rules don't work with docker images (which have several layers and files). That's why `DeleteDockerImagesOlderThan` rule exists in addition to `DeleteOlderThan`.

**Comment by donhui** [comment]
I have defined the following rules, which can meet my needs. You can refer to them.
```
CleanupPolicy(
    'docker-demo-cleanup',
    rules.repo('docker-demo'),
    rules.exclude_docker_images(['*:latest', '*:release*']),
    rules.delete_docker_images_older_than(days=30),
    rules.exclude_path("base-tools*"),
    rules.keep_latest_n_file_in_folder(count=3),
)
```

**Comment by allburov** [comment]
@tahajahangir do you have any luck with it?
If it works - we can add as a reference to FAQ section

**Comment by tahajahangir** [comment]
Yes, It works correctly

**Comment by tahajahangir** [comment]
Yes, it works

**Comment by afolgado** [comment]
As I can see, it won't work. It won't keep N tags per image, but N files per tag. In Docker images, every tag is a folder and every file in that folder is a layer of the image (plus the manifes.json file). This will delete some of the layers of every non-excluded tag, making them useless.

**Comment by seth-stvns** [comment]
I am seeing the same behavior as @afolgado. It is deleting a few layers from every image. Based on artifactory's layout of docker images and other resources (maven, gradle, npm, etc) this filter should be checking the folder directly above the files.

Will dig more into the code to see if I can find the issue when I have some time.

**Comment by donhui** [comment]
> I have defined the following rules, which can meet my needs. You can refer to them.
> 
> ```
> CleanupPolicy(
>     'docker-demo-cleanup',
>     rules.repo('docker-demo'),
>     rules.exclude_docker_images(['*:latest', '*:release*']),
>     rules.delete_docker_images_older_than(days=30),
>     rules.exclude_path("base-tools*"),
>     rules.keep_latest_n_file_in_folder(count=3),
> )
> ```

In this case,  the rule `rules.delete_docker_images_older_than(days=30)` is required.
It can filter some docker images to delete.
Then the rule `rules.keep_latest_n_file_in_folder(count=3)` will filter on this basis.

**Comment by zbialik** [comment]
FWIW this feature request specifically asks for a method to keep latest by download time, not creation time. Because of this, I don't believe `KeepLatestNFilesInFolder` achieves the desired request. 

I'm now trying to write a custom rule that's basically the same as `KeepLatestNFilesInFolder` except that it's based on the download time, rather than the creation time. Not sure if anyone else has implemented something similar.

**Comment by allburov** [comment]
I see, so like `KeepLatestNFiles` but for docker? 
I'm not sure, but may be `KeepLatestNFiles` could work with docker rules as well, could one try it out?

**Comment by allburov** [comment]
Sorry, I see the solution for that was accepted 
https://github.com/devopshq/artifactory-cleanup/issues/61#issuecomment-1231719136

**Comment by allburov** [comment]
@zbialik please share with us the solution when it's ready, I'll be happy to review and accept a PR! :pray:
